Filter by list

hayesgb opened the pull request:
- [ X ] Closes #9411 
- [ X ] Tests added / passed
- [ X ] Passes `pre-commit run --all-files`

This PR addresses the fact that filtering a Dask DataFrame with `ddf.isin(L)` if L is a large list, is much slower than if L is a numpy array.  This appears to be a result of serializing lists with pickle, while serializing a numpy array uses dask's custom serializer.

Here we do a few things to reduce the serialization cost if L is a list:

1. If L is a list, we remove duplicate
2. If all the elements in L are of the same datatype, we convert L to a numpy array, otherwise we are forced to incur the serialization cost

GPUtester commented:
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

pavithraes commented:
add to allowlist

ian-r-rose reviewed (commented):
Thanks @hayesgb, this seems like a reasonable workaround, but I was worried enough about the poor performance that I looked into it a bit more.

Serialization of long python collections isn't just slow, it's pathologically slow, and takes a huge amount of CPU. (When @pavithraes ran the reproducer, she saw a huge amount of garbage collection warnings). And it must be related to serialization, since no such behavior happens for a single-machine scheduler:

```python
import numpy as np
import distributed
import dask
import time

values = list(np.random.randint(2**30, size=1_000_000))

start = time.time()
dask.delayed(values, traverse=False).compute(scheduler="threads")
end = time.time()
print(end-start)  # 0.01 seconds
```
But with a distributed scheduler:

```python
import numpy as np
import distributed
import dask
import time

cluster = distributed.LocalCluster()
client = distributed.Client(cluster)

values = [1] * 1_000_000

start = time.time()
dask.delayed(values, traverse=False).compute()
end = time.time()
print(end-start)  # 37 seconds!
```

Furthermore, this isn't *just* due to `pickle.dumps` of the python objects. Serializing them directly with `pickle.dumps` just takes a fraction of a second:

```python
pickle.dumps(values)  # 0.026 seconds
```

I'm fairly certain that we are hitting https://github.com/dask/distributed/issues/6368 . In particular, the serialization framework isn't just serializing the list of values, but traversing that list (unnecessarily, IMO), and serializing *each value* independently. And if my list has things like `np.int64` objects in it instead of native python `int`s, things get slower by another factor of ~2!

So I think something along these lines is a good fix, but I also think that we might want to take a closer look at the serialization logic to see what is going so wrong.

ian-r-rose commented:
One workflow (seemingly untested) that this PR would break is if the values by which we are filtering are themselves lazy dask objects (so, if the thing we want to filter by is the result of some computation). The following works on `main`, but fails on this branch:

```python
import dask.datasets

ddf = dask.datasets.timeseries()

ddf[ddf.id.isin([dask.delayed(1000), dask.delayed(1001)])].compute()
```

Now, this is a pretty niche usage, but it does motivate me to look a little harder at the serialization issue as the root cause.

hayesgb commented:
Thanks @ian-r-rose.  I didn't realize we were always recursing into lists during serialization.  If we can get that fixed, it would be great.

I've added a small update for consideration if the change to serialization process looks like its going to be a heavy lift.

hayesgb commented:
@ian-r-rose -- I added a few additional tests.  There's one based on your:

```ddf[ddf.id.isin([dask.delayed(1000), dask.delayed(1001)])].compute()```

from above, which also fails on main.

hayesgb commented:
@ian-r-rose -- Digging into this a little more.  Given:

```
dsk = {
    ("x", 0): pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]}, index=[0, 1, 3]),
    ("x", 1): pd.DataFrame({"a": [4, 5, 6], "b": [3, 2, 1]}, index=[5, 6, 8]),
    ("x", 2): pd.DataFrame({"a": [7, 8, 9], "b": [0, 0, 0]}, index=[9, 9, 9]),
}
meta = make_meta(
    {"a": "i8", "b": "i8"}, index=pd.Index([], "i8"), parent_meta=pd.DataFrame()
)
d = dd.DataFrame(dsk, "x", meta, [0, 5, 9, 9])
full = d.compute()
f_list_delayed = [delayed(1), delayed(2), delayed(3)]
```

On `main` your example from above computes successfully:
`ddf.a.isin(f_list_delayed).compute()`

While this fails:
`assert_eq(d.a.isin(f_list_delayed), full.a.isin(f_list_delayed))`

But this, which seems more idiomatic to me, executes successfully:
`assert_eq(d.a.isin([d.compute() for d in f_list_delayed]), full.a.isin([d.compute() for d in f_list_delayed]))`

If we accept that the pandas assertion should pass (which is consistent with main) we can remove the `is_dask_collection` check.

hayesgb commented:
Discussed the above with @jrbourbeau and aligned on:
`assert_eq(ddf.a.isin(f_list_delayed), df.a.isin(f_list))`

Also added a test to confirm we handle filtering by a list of lists, preferring numpy array for this operation, when it it supported, otherwise fall back to recursive iteration.

Updated the reproducer.py to:
```
import time
import numpy as np
import pandas as pd
from dask.dataframe import from_pandas
from distributed import LocalCluster, Client
import copy
import sys


def main():
    pd_list_result, pd_array_result, dd_list_result, dd_array_result = [], [], [], []
    pd_list_length, pd_array_length, dd_list_length, dd_array_length = [], [], [], []

    n = 10_000_000
    rs = np.random.RandomState(42)
    a_column_unique_values = np.arange(1, n // 10)
    df = pd.DataFrame({"A": rs.choice(a_column_unique_values, n), "B": rs.random(n)})

    filter_values_list = sorted(
        rs.choice(a_column_unique_values, len(a_column_unique_values) // 2).tolist()
    )
    filter_values_array = np.array(filter_values_list, dtype="i8")

    for i in list(range(5)):
        t1 = time.time()
        tmp_df = df[df["A"].isin(filter_values_list)].copy()
        len_after_filter = len(tmp_df)
        pd_list_result.append(time.time() - t1)
        pd_list_length.append(len_after_filter)

        t1 = time.time()
        tmp_df = df[df["A"].isin(filter_values_array)].copy()
        len_after_filter = len(tmp_df)
        pd_array_result.append(time.time() - t1)
        pd_array_length.append(len_after_filter)

        with LocalCluster(n_workers=4, processes=True, threads_per_worker=1) as local_cluster:
            with Client(local_cluster) as client:
                ddf = from_pandas(df, npartitions=8).persist()

                t1 = time.time()
                tmp_ddf = ddf[ddf["A"].isin(filter_values_list)].copy()
                len_after_filter = len(tmp_ddf)
                t2 = time.time()
                dd_list_result.append(t2-t1)
                dd_list_length.append(len_after_filter)

                t1 = time.time()
                tmp_ddf = ddf[ddf["A"].isin(filter_values_array)].copy()
                len_after_filter = len(tmp_ddf)
                t2 = time.time()
                dd_array_result.append(t2-t1)
                dd_array_length.append(len_after_filter)

    pd_list_length = list(set(pd_list_length))
    pd_array_length = list(set(pd_array_length))
    dd_list_length = list(set(dd_list_length))
    dd_array_length = list(set(dd_array_length))
    assert all(len(s)==1 for s in [pd_list_length, pd_array_length, dd_list_length, dd_array_length])
    print(f"Mean time of 5 runs to filter Pandas DataFrame with list:  {np.mean(pd_list_result):,.3f}s, with output len:  {pd_list_length[0]}")
    print(f"Mean time of 5 runs to filter Pandas DataFrame with array:  {np.mean(pd_array_result):,.3f}s, with output len:  {pd_array_length[0]}")
    print(f"Mean time of 5 runs filter Dask Dataframe with list:  {np.mean(dd_list_result):,.3f}s, with output len:  {dd_list_length[0]}")
    print(f"Mean time of 5 runs filter Dask DataFrame with array:  {np.mean(dd_array_result):,.3f}s, with output len:  {dd_array_length[0]}")
```

Running this on `main` I get:
![Screen Shot 2022-09-06 at 12 56 45 PM](https://user-images.githubusercontent.com/12595382/188705960-8927fed4-82c9-4713-a5af-78d16486ca57.png)


Running against this branch, I get:
![Screen Shot 2022-09-06 at 12 53 29 PM](https://user-images.githubusercontent.com/12595382/188705422-0d7bdabe-cb79-42e5-9214-7f8a64050f88.png)



Of note -- this PR reduces execution time on my machine from an average of 25.2 seconds to 1.61 seconds over 5 runs.

ian-r-rose commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -50,6 +51,18 @@
 d = dd.DataFrame(dsk, "x", meta, [0, 5, 9, 9])
 full = d.compute()
+
+dsk2 = {
+    ("x", 0): pd.DataFrame({"a": [[1, 2, 3]], "b": [4, 5, 6]}, index=[0, 1, 3]),
+    ("x", 1): pd.DataFrame({"a": [4, 5, 6], "b": [[3, 4, 5]]}, index=[5, 6, 8]),
+    ("x", 2): pd.DataFrame({"a": [7, 8, 9], "b": [0, 0, 0]}, index=[9, 9, 9]),
+}
+meta2 = make_meta(
+    {"a": "i8", "b": "i8"}, index=pd.Index([], "i8"), parent_meta=pd.DataFrame()
+)
+d2 = dd.DataFrame(dsk2, "x", meta2, [0, 5, 9, 9])
+full2 = d2.compute()
```
This looks identical to the above dataframe `d`, and is unused below. Is there a reason to make another?

A meta point -- the above snippet constructing a dataframe looks quite old, and for new tests I would probably advocate for using `dd.from_pandas()` rather than doing manual construction.

hayesgb replied:
Done.

ian-r-rose commented:
We had previously chatted offline about using `dask.base.unpack_collections` to identify possibly-nested dask collections instead of using `any(is_dask_collection(d) ...)`. Any additional thoughts about that?

hayesgb commented:
This should be a less expensive operation in the case when  `is_dask_collection` is `True`, and as a generator, should be more memory efficient.  By contrast `unpack_collections` requires iterating through the entire list regardless of its contents, unpacking and repacking the collections, and materializes both results.

ian-r-rose reviewed (approved):
> This should be a less expensive operation in the case when `is_dask_collection` is `True`, and as a generator, should be more memory efficient. By contrast `unpack_collections` requires iterating through the entire list regardless of its contents, unpacking and repacking the collections, and materializes both results.

Makes sense. `unpack_collections` would handle the case where there is a *nested* collection, which this approach would not, but that use-case is so niche that I think I would be comfortable moving forward with this.

ian-r-rose commented:
Thanks @hayesgb!
